File: 0022-generate-parentheses/0022-generate-parentheses.java
class Solution {
    public List<String> generateParenthesis(int n) {
        if(n==1) return new ArrayList<String>(Arrays.asList("()"));
        Set<String> set=new HashSet<String>();
        for(String str:generateParenthesis(n-1)){
            for(int i=0;i<str.length();i++){
                set.add(str.substring(0,i+1)+"()"+str.substring(i+1,str.length()));               
            }            
        }
        List<String> list = new ArrayList<String>(set);
        return list;
    }
}

File: 0022-generate-parentheses/0022-generate-parentheses.cpp
class Solution {
public:
    void f(int n, int l, int r, string paren, vector<string> &ans){
        if(l==n && r==l) ans.push_back(paren);
        else{
            if(l<n){
                f(n,l+1, r, paren + '(', ans);
            } if(r<l){
                f(n,l, r+1, paren + ')', ans);
            }
        }
    }
    vector<string> generateParenthesis(int n) {
        vector<string> ans;
        string paren = "";

        f(n,0,0,paren, ans);
        return ans;
    }
};